### ISSUE TITLE
[BUG]  Books Not Sorting Correctly in Stump

### ISSUE THREAD

**Issue by Ammoton** [stumpapp/stump#482]
## Description of the Bug

Books that are named Volume 1, Volume 2, Volume 3, ... are sorted incorrectly.

They appear in this order:

    1
    10
    11
    ...
    19
    2
    20

## Expected Behavior

They should be sorted like this:

    1
    2
    3
    ...
    9
    10
    11

**Environment:**

- OS: [Windows]

**Build Details:**

- Version: [0.0.7]
- Docker: [no]

**Comment by aaronleopold** [comment]
> Books that are named Volume 1, Volume 2, Volume 3, ... are sorted incorrectly.

When you say named, are you referring to the filename (e.g. `Volume 1.zip`) or the name pulled from metadata?

**Comment by Ammoton** [comment]
I'm referring to the filename. But the title in the metadata is also set. I tested a little more, and if I only use the filename without metadata, it's not sorted correctly. If I use metadata (title) and filenames like 01, 02, etc., it works. But if I use 1, 2, etc., it doesn't. Could it be that the filename is prioritized?

**Comment by aaronleopold** [comment]
Yeah IIRC the filename is most likely what is being prioritized at the moment. Coincidentally I started a refactor for the sorting implementation a week or so ago [here](https://github.com/stumpapp/stump/tree/experiment/pin-prisma). I think I'll try to explore fixing this issue there, once what I have is viable, as it makes ordering by relations (i.e. the metadata) a more possible.

I'll report back once I have something. Thanks for letting me know!

**Comment by aaronleopold** [comment]
I wanted to post a little update just to say I haven't forgotten about this. I have a viable fix working, but there is a lot of UI work associated with the overhaul to how ordering and filtering works. My progress is contained to this branch for any code-savvy followers: https://github.com/stumpapp/stump/tree/experiment/pin-prisma. I'll aim to get it into experimental by the end of December, ideally

**Comment by aaronleopold** [comment]
> I'll aim to get it into experimental by the end of December, ideally

That didn't happen 😓 

I promise this hasn't slipped my radar though. I'll be rounding out granular scans first, other smaller-scoped bug fixes, and then plan to revisit that in progress branch with _potential_ fixes. The big headache is potentially having to make upstream changes to the prisma client, but I intend to eventually get around to fixing this.

**Comment by aaronleopold** [comment]
The current plan to solve this is to add two mechanisms for controlling the default sort:

1. A server-wide default, stored in `server_config`
2. A per-library default, stored in `library_configs`

My idea is to just use one of those two values depending on the context and whether a client requested a specific sort. For example:

- Querying all books on server -> server-wide default
- Query books scoped within a specific library/series -> library's default OR server-wide default

This will not automatically solve the natural order sort problem, however it would allow you to sort based on e.g. `{ field: MediaMetadata::Position, direction: Direction::Asc }`. You should be able to pad any positions in the name to work around this in the meantime, e.g. `Volume 001`, `Volume 002`, etc.
